OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,338 @@
+            {
+                return;
+            }
+            _lastUpdate = Time.realtimeSinceStartup;
+            long hash = 0;
+            var keysCount = 0;
+            var countdownTrigger = true;
+            var keys = new int[MaxComboLength];
+            for (int code = MinUsefulKey; code < MaxUsefulKey; code++)
+            {
+                if (!(Enum.IsDefined(typeof(KeyCode), (KeyCode)code) && UnityEngine.Input.GetKey((KeyCode)code)))
+                {
+                    continue;
+                }
+                keys[keysCount] = code;
+                keysCount++;
+                if (keysCount > MaxComboLength)
+                {
+                    hash = (int)RegistrationCode.CombinationTooLong;
+                    break;
+                }
+                hash = hash * MaxUsefulKey + code;
+                if (Time.realtimeSinceStartup - _timeout[code] > Cooldown)
+                {
+                    countdownTrigger = false;
+                }
+            }
+            if (!_comboRegistry.ContainsKey(hash))
+            {
+                if (_lastPressed != null)
+                {
+                    _lastPressed.SetPressed(false);
+                }
+                _lastPressed = null;
+                return;
+            }
+            var combination = _comboRegistry[hash];
+            if (!(combination == _lastPressed) && countdownTrigger)
+            {
+                return;
+            }
+            if (_lastPressed != null && _lastPressed != combination)
+            {
+                _lastPressed.SetPressed(false);
+            }
+            _lastPressed = (ModInputCombination)combination;
+            _lastPressed.SetPressed();
+            if (keysCount > 1)
+            {
+                return;
+            }
+            for (var i = 0; i < keysCount; i++)
+            {
+                _timeout[keys[i]] = Time.realtimeSinceStartup;
+            }
+        }
+
+        public bool IsPressedExact(IModInputCombination combination)
+        {
+            UpdateCombo();
+            return _lastPressed == combination;
+        }
+
+        public bool IsNewlyPressedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCombo();
+            return _lastPressed == combination && combination.IsFirst(keep);
+        }
+
+        public bool WasTappedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCombo();
+            return combination != _lastPressed
+                && (combination.PressDuration < TapDuration)
+                && combination.IsFirst(keep);
+        }
+
+        public bool WasNewlyReleasedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCombo();
+            return _lastPressed != combination && combination.IsFirst(keep);
+        }
+
+        private bool IsPressedSingle(IModInputCombination combination)
```

Comment:
This method is a bit complicated to read. I think it would be nice if you extract the contents into smaller methods with intuitive names.

Reply:
separated part that updates _singlePressed